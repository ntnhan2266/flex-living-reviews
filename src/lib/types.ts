export interface HostawayReview {
  id: number;
  type: string;
  status: string;
  rating: number | null;
  publicReview: string;
  reviewCategory: {
    category: string;
    rating: number;
  }[];
  submittedAt: string;
  guestName: string;
  listingName: string;
}

export type Channel = 'hostaway' | 'airbnb' | 'booking' | 'google';
export type ReviewStatus =
  | 'awaiting'
  | 'pending'
  | 'scheduled'
  | 'submitted'
  | 'published'
  | 'expired'
  | 'approved'
  | 'public';

export interface NormalizedReview {
  id: string;
  propertyName: string;
  guestName: string;
  rating: number;
  comment: string;
  date: Date;
  channel: Channel;
  categories: Record<string, number | undefined>;
  status: ReviewStatus;
}

export interface PropertyStats {
  id: string;
  name: string;
  averageRating: number;
  totalReviews: number;
  approvedReviews: number;
  recentTrend: 'up' | 'down' | 'stable';
}

export type CategoryKey =
  | 'cleanliness'
  | 'communication'
  | 'location'
  | 'accuracy'
  | 'checkin'
  | 'value';

export type SortKey = 'date_desc' | 'date_asc' | 'rating_desc' | 'rating_asc';

export type Filters = {
  property: string;
  channel: '' | 'hostaway' | 'airbnb' | 'booking' | 'google';
  rating: '' | '1' | '2' | '3' | '4' | '5';
  startDate: string; // yyyy-mm-dd
  endDate: string;   // yyyy-mm-dd
  status: 'all' | 'pending' | 'approved';
  category?: '' | CategoryKey;
  minCategory?: '' | '1' | '2' | '3' | '4' | '5';
  sort?: SortKey;
};

export interface ApiResponse<T> {
  status: 'success' | 'error';
  data?: T;
  message?: string;
  total?: number;
}
