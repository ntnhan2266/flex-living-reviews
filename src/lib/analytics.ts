// src/lib/analytics.ts
import { NormalizedReview, ReviewStatus } from '@/lib/types';

export type Granularity = 'day' | 'week' | 'month';

export type TimeseriesDatum = {
  date: string; // "YYYY-MM-DD" for day/week start/month start
  count: number; // number of reviews in bucket
  avgRating: number; // 0..5 average in bucket (0 when no ratings)
};

/**
 * Buckets a JS Date into an ISO key by granularity:
 * - day  -> YYYY-MM-DD
 * - week -> YYYY-MM-DD of the week's Monday
 * - month-> YYYY-MM-01
 */
export function bucketKey(d: Date, granularity: Granularity): string {
  const dt = new Date(d);
  dt.setHours(0, 0, 0, 0);

  if (granularity === 'week') {
    // Monday as week start
    const day = dt.getDay() || 7; // 1..7 (Mon..Sun)
    if (day !== 1) dt.setDate(dt.getDate() - (day - 1));
  } else if (granularity === 'month') {
    dt.setDate(1);
  }

  // fastest way to get YYYY-MM-DD without deps
  return dt.toISOString().slice(0, 10);
}

/**
 * Build a timeseries for reviews.
 * @param reviews - list of reviews
 * @param options.statuses - optional statuses to include (e.g. ['approved'])
 * @param options.granularity - day|week|month (default 'day')
 */
export function getReviewTimeseries(
  reviews: NormalizedReview[],
  options?: { statuses?: ReviewStatus[]; granularity?: Granularity },
): TimeseriesDatum[] {
  const allowed = new Set(options?.statuses ?? []);
  const gran = options?.granularity ?? 'day';

  const grouped: Record<string, { count: number; sum: number; withRating: number }> = {};

  for (const r of reviews) {
    if (allowed.size && !allowed.has(r.status)) continue;

    // normalize date (handle string | Date)
    const d = r.date instanceof Date ? r.date : new Date(r.date);
    const key = bucketKey(d, gran);
    if (!grouped[key]) grouped[key] = { count: 0, sum: 0, withRating: 0 };

    grouped[key].count += 1;

    const val = Number(r.rating);
    if (Number.isFinite(val)) {
      grouped[key].sum += val;
      grouped[key].withRating += 1;
    }
  }

  return Object.entries(grouped)
    .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
    .map(([date, { count, sum, withRating }]) => ({
      date,
      count,
      avgRating: withRating ? Number((sum / withRating).toFixed(2)) : 0,
    }));
}
