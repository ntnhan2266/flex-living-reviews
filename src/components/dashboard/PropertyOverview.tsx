import React from 'react';
import { PropertyStats } from '@/lib/types';
import { Minus, MoveDownRight, MoveUpRight } from 'lucide-react';

interface PropertyOverviewProps {
  properties: PropertyStats[];
}

const PropertyOverview: React.FC<PropertyOverviewProps> = ({ properties }) => {
  if (!properties || properties.length === 0) {
    return (
      <div className="bg-white p-6 rounded-lg shadow text-center text-gray-900">
        No property data available.
      </div>
    );
  }

  return (
    <div className="bg-white p-6 rounded-lg shadow">
      <h2 className="text-xl font-semibold mb-4 text-gray-800">Property Overview</h2>
      <div className="overflow-x-auto">
        <table className="min-w-full divide-y divide-gray-200">
          <thead>
            <tr>
              <th
                scope="col"
                className="px-4 py-2 text-left text-xs font-medium text-gray-900 uppercase tracking-wider"
              >
                Property
              </th>
              <th
                scope="col"
                className="px-4 py-2 text-right text-xs font-medium text-gray-900 uppercase tracking-wider"
              >
                Total Reviews
              </th>
              <th
                scope="col"
                className="px-4 py-2 text-right text-xs font-medium text-gray-900 uppercase tracking-wider"
              >
                Avg Rating
              </th>
              <th
                scope="col"
                className="px-4 py-2 text-right text-xs font-medium text-gray-900 uppercase tracking-wider"
              >
                Approved
              </th>
              <th
                scope="col"
                className="px-4 py-2 text-right text-xs font-medium text-gray-900 uppercase tracking-wider"
              >
                Trend
              </th>
            </tr>
          </thead>
          <tbody className="bg-white divide-y divide-gray-100">
            {properties.map((p) => (
              <tr key={p.id}>
                <td className="px-4 py-2 whitespace-nowrap font-medium text-gray-900">
                  <a href={`/property/${p.id}`} className="text-blue-600 hover:underline">
                    {p.name}
                  </a>
                </td>
                <td className="px-4 py-2 whitespace-nowrap text-right text-gray-900">
                  {p.totalReviews}
                </td>
                <td className="px-4 py-2 whitespace-nowrap text-right text-gray-900">
                  {p.averageRating != null ? p.averageRating.toFixed(1) : '-'}
                </td>
                <td className="px-4 py-2 whitespace-nowrap text-right text-gray-900">
                  {p.approvedReviews ?? 0}
                </td>
                <td className="px-4 py-2 whitespace-nowrap text-right text-gray-900 flex items-center justify-end">
                  {p.recentTrend === 'stable' && <Minus color="#9CA3AF" />}
                  {p.recentTrend === 'up' && <MoveUpRight color="#01782c" />}
                  {p.recentTrend === 'down' && <MoveDownRight color="#ff0000" />}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
};

export default PropertyOverview;
