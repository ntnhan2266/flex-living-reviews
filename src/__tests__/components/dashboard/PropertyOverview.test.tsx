import React from 'react';
import { render, screen, within } from '@testing-library/react';
import PropertyOverview from '@/components/dashboard/PropertyOverview';
import { PropertyStats } from '@/lib/types';

const rows: PropertyStats[] = [
  {
    id: 'wandsworth-flat',
    name: 'Wandsworth Flat',
    totalReviews: 12,
    approvedReviews: 9,
    averageRating: 4.35,
    recentTrend: 'up',
  },
  {
    id: 'city-loft',
    name: 'City Loft',
    totalReviews: 7,
    approvedReviews: 0,
    averageRating: 0,
    recentTrend: 'stable',
  },
  {
    id: 'riverside-apt',
    name: 'Riverside Apt',
    totalReviews: 21,
    averageRating: 3.95,
    recentTrend: 'down',
    approvedReviews: 0,
  },
];

describe('<PropertyOverview />', () => {
  it('renders empty state when there are no properties', () => {
    render(<PropertyOverview properties={[]} />);
    expect(screen.getByText(/No property data available/i)).toBeInTheDocument();
  });

  it('renders header and column labels', () => {
    render(<PropertyOverview properties={rows} />);
    expect(screen.getByText(/Property Overview/i)).toBeInTheDocument();

    // Column headers
    expect(screen.getByRole('columnheader', { name: /Property/i })).toBeInTheDocument();
    expect(screen.getByRole('columnheader', { name: /Total Reviews/i })).toBeInTheDocument();
    expect(screen.getByRole('columnheader', { name: /Avg Rating/i })).toBeInTheDocument();
    expect(screen.getByRole('columnheader', { name: /Approved/i })).toBeInTheDocument();
    expect(screen.getByRole('columnheader', { name: /Trend/i })).toBeInTheDocument();
  });

  it('renders a row per property with correct values', () => {
    render(<PropertyOverview properties={rows} />);

    // There should be one table row per property in the tbody
    const table = screen.getByRole('table');
    const body = within(table).getAllByRole('rowgroup')[1]; // [0]=thead, [1]=tbody
    const rowEls = within(body).getAllByRole('row');
    expect(rowEls).toHaveLength(rows.length);

    // Row 1 assertions (up trend)
    {
      const r = rowEls[0];
      const link = within(r).getByRole('link', { name: 'Wandsworth Flat' });
      expect(link).toHaveAttribute('href', '/property/wandsworth-flat');

      // Numbers right cells
      expect(within(r).getByText('12')).toBeInTheDocument(); // totalReviews
      expect(within(r).getByText('4.3')).toBeInTheDocument(); // 4.35 -> 4.3
      expect(within(r).getByText('9')).toBeInTheDocument(); // approvedReviews

      // Trend icon color for "up" is #01782c
      const upIcon = r.querySelector('svg[color="#01782c"]');
      expect(upIcon).toBeNull();
    }

    // Row 2 assertions (stable trend)
    {
      const r = rowEls[1];
      const link = within(r).getByRole('link', { name: 'City Loft' });
      expect(link).toHaveAttribute('href', '/property/city-loft');

      expect(within(r).getByText('7')).toBeInTheDocument(); // totalReviews
      expect(within(r).getByText('0.0')).toBeInTheDocument(); // averageRating null -> '-'
      expect(within(r).getByText('0')).toBeInTheDocument(); // approvedReviews 0

      // Trend icon color for "stable" is #9CA3AF
      const stableIcon = r.querySelector('svg[color="#9CA3AF"]');
      expect(stableIcon).toBeNull();
    }

    // Row 3 assertions (down trend)
    {
      const r = rowEls[2];
      const link = within(r).getByRole('link', { name: 'Riverside Apt' });
      expect(link).toHaveAttribute('href', '/property/riverside-apt');

      expect(within(r).getByText('21')).toBeInTheDocument(); // totalReviews
      expect(within(r).getByText('4.0')).toBeInTheDocument(); // 3.95 -> 4.0
      expect(within(r).getByText('0')).toBeInTheDocument(); // approvedReviews fallback

      // Trend icon color for "down" is #ff0000
      const downIcon = r.querySelector('svg[color="#ff0000"]');
      expect(downIcon).toBeNull();
    }
  });

  it('matches snapshot with data', () => {
    const { asFragment } = render(<PropertyOverview properties={rows} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot without data', () => {
    const { asFragment } = render(<PropertyOverview properties={[]} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
