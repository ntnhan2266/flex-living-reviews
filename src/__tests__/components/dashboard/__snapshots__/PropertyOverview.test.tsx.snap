// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<PropertyOverview /> matches snapshot with data 1`] = `
<DocumentFragment>
  <div
    class="bg-white p-6 rounded-lg shadow"
  >
    <h2
      class="text-xl font-semibold mb-4 text-gray-800"
    >
      Property Overview
    </h2>
    <div
      class="overflow-x-auto"
    >
      <table
        class="min-w-full divide-y divide-gray-200"
      >
        <thead>
          <tr>
            <th
              class="px-4 py-2 text-left text-xs font-medium text-gray-900 uppercase tracking-wider"
              scope="col"
            >
              Property
            </th>
            <th
              class="px-4 py-2 text-right text-xs font-medium text-gray-900 uppercase tracking-wider"
              scope="col"
            >
              Total Reviews
            </th>
            <th
              class="px-4 py-2 text-right text-xs font-medium text-gray-900 uppercase tracking-wider"
              scope="col"
            >
              Avg Rating
            </th>
            <th
              class="px-4 py-2 text-right text-xs font-medium text-gray-900 uppercase tracking-wider"
              scope="col"
            >
              Approved
            </th>
            <th
              class="px-4 py-2 text-right text-xs font-medium text-gray-900 uppercase tracking-wider"
              scope="col"
            >
              Trend
            </th>
          </tr>
        </thead>
        <tbody
          class="bg-white divide-y divide-gray-100"
        >
          <tr>
            <td
              class="px-4 py-2 whitespace-nowrap font-medium text-gray-900"
            >
              <a
                class="text-blue-600 hover:underline"
                href="/property/wandsworth-flat"
              >
                Wandsworth Flat
              </a>
            </td>
            <td
              class="px-4 py-2 whitespace-nowrap text-right text-gray-900"
            >
              12
            </td>
            <td
              class="px-4 py-2 whitespace-nowrap text-right text-gray-900"
            >
              4.3
            </td>
            <td
              class="px-4 py-2 whitespace-nowrap text-right text-gray-900"
            >
              9
            </td>
            <td
              class="px-4 py-2 whitespace-nowrap text-right text-gray-900 flex items-center justify-end"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-move-up-right"
                fill="none"
                height="24"
                stroke="#01782c"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13 5H19V11"
                />
                <path
                  d="M19 5L5 19"
                />
              </svg>
            </td>
          </tr>
          <tr>
            <td
              class="px-4 py-2 whitespace-nowrap font-medium text-gray-900"
            >
              <a
                class="text-blue-600 hover:underline"
                href="/property/city-loft"
              >
                City Loft
              </a>
            </td>
            <td
              class="px-4 py-2 whitespace-nowrap text-right text-gray-900"
            >
              7
            </td>
            <td
              class="px-4 py-2 whitespace-nowrap text-right text-gray-900"
            >
              0.0
            </td>
            <td
              class="px-4 py-2 whitespace-nowrap text-right text-gray-900"
            >
              0
            </td>
            <td
              class="px-4 py-2 whitespace-nowrap text-right text-gray-900 flex items-center justify-end"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-minus"
                fill="none"
                height="24"
                stroke="#9CA3AF"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 12h14"
                />
              </svg>
            </td>
          </tr>
          <tr>
            <td
              class="px-4 py-2 whitespace-nowrap font-medium text-gray-900"
            >
              <a
                class="text-blue-600 hover:underline"
                href="/property/riverside-apt"
              >
                Riverside Apt
              </a>
            </td>
            <td
              class="px-4 py-2 whitespace-nowrap text-right text-gray-900"
            >
              21
            </td>
            <td
              class="px-4 py-2 whitespace-nowrap text-right text-gray-900"
            >
              4.0
            </td>
            <td
              class="px-4 py-2 whitespace-nowrap text-right text-gray-900"
            >
              0
            </td>
            <td
              class="px-4 py-2 whitespace-nowrap text-right text-gray-900 flex items-center justify-end"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-move-down-right"
                fill="none"
                height="24"
                stroke="#ff0000"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 13V19H13"
                />
                <path
                  d="M5 5L19 19"
                />
              </svg>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<PropertyOverview /> matches snapshot without data 1`] = `
<DocumentFragment>
  <div
    class="bg-white p-6 rounded-lg shadow text-center text-gray-900"
  >
    No property data available.
  </div>
</DocumentFragment>
`;
